File: tools/gz_cloc.py
#!/usr/bin/python

import subprocess
import datetime

# Warning: This script will update the current repository!!

# How to use
# 1. This run this script in the root director of Gazebo.
# 2. The output will be written to /tmp/gz_loc.csv
# 3. Load into a Google spreadsheet.
# 4. Generate a graph, and select Time Line as the graph type.

start_year = 2007

f = open('/tmp/gz_loc.csv', 'w')

now = datetime.datetime.now()

f.write("date,blank,comment,code\n")

for y in range(start_year, now.year+1):
  if y == now.year:
    end_month = now.month + 1
  else:
    end_month = 13

  for m in range(1,end_month):
    date = str(y) + "-" + str(m) 
    cmd = "hg up -d \"" + date + "\""
    print cmd
    os.system(cmd)

    cmd = "cloc --force-lang=\"C++\",cc --force-lang=\"C++\",c --force-lang=\"C++\",hh --force-lang=\"C++\",h --force-lang=\"C++\",hpp --exclude-lang=\"XML\",\"HTML\",\"XSD\",\"Python\",\"make\",\"Bourne Shell\",\"Javascript\",\"CSS\",\"m4\",\"Ruby\",\"DOS Batch\" --exclude_dir=deps,Media,media,cmake,doc,build --csv --quiet --progress-rate=0 * | tail -n 1"
    proc = subprocess.Popen([cmd], stdout=subprocess.PIPE, shell=True)
    (out, err) = proc.communicate()
    out_parts = out.split(',')
    f.write("%d/01/%d, %s, %s, %s" % (m, y, out_parts[2], out_parts[3], out_parts[4]))

f.close()

File: tools/gz_cloc.rb
#!/usr/bin/env ruby

# Warning: This script will update the current repository!!
#
# How to use
# 1. Run this script in the root directory of Gazebo.
# 2. The output will be written to /tmp/gz_loc.csv
# 3. Load into a Google spreadsheet.
# 4. Generate a graph, and select Time Line as the graph type.

printf("Warning: This will modify the current repository.\n")
printf("Do you want to continue (Y/n): ")
char = gets

if char != "Y\n"
  abort
end

startYear = 2007

f = File.open('/tmp/gz_loc.csv', 'w')

time = Time.new

f.printf("date,blank,comment,code\n")

# Loop through all the years.
for y in (startYear..time.year)

  # Get the end month
  endMonth = y == time.year ? time.month : 12

  # Loop through all the months.
  for m in (1..endMonth)
    date = y.to_s + "-" + m.to_s

    # Update the repo
    `hg up -C -d "#{date}"`

    # Count lines of code
    result = `cloc --force-lang=\"C++\",cc --force-lang=\"C++\",c --force-lang=\"C++\",hh --force-lang=\"C++\",h --force-lang=\"C++\",hpp --exclude_dir=deps,Media,media,cmake,doc,build --csv --quiet --progress-rate=0 *`

    lines = result.split("\n")

    # Grab just C++ results
    lines.each do |line|
      if line.include?("C++")
        parts = line.split(",")
        printf("%d/01/%d, %s, %s, %s\n", m, y, parts[2], parts[3], parts[4])
        f.printf("%d/01/%d, %s, %s, %s\n", m, y, parts[2], parts[3], parts[4])
      end
    end
  end
end

f.close()
